s3torchconnector/src/s3torchconnector/s3iterable_dataset.py

```diff
@@ -133,6 +148,47 @@ def _get_transformed_object(self, bucket_key: S3BucketKeyData) -> Any:
         )
 
     def __iter__(self) -> Iterator[Any]:
+        worker_id = 0
+        num_workers = 1
+        if self._enable_sharding:
+            worker_info = torch.utils.data.get_worker_info()
+            if worker_info is not None:
+                worker_id = worker_info.id
+                num_workers = worker_info.num_workers
+
+        """"
+        In a multi-process setting (e.g., distributed training), the dataset needs to be
+        sharded across multiple processes. The following variables control this sharding:
+
+        _rank: The rank (index) of the current process within the world (group of processes).
+        _world_size: The total number of processes in the world (group).
+
+        In addition, within each process, the dataset may be further sharded across multiple
+        worker threads or processes (e.g., for data loading). The following variables control
+        this intra-process sharding:
+
+        worker_id: The ID of the current worker thread/process within the process.
+        num_workers: The total number of worker threads/processes within the process.
+
+        The _shard_index and _shard_count variables are computed based on the above values,
+        and they determine which subset of the dataset objects should be processed by the
+        current worker thread/process in the current process rank.
+        """
+
+        self._shard_index = num_workers * self._rank + worker_id
```

Comment:
are we assuming that number of workers per rank would be equal everytime? If the number of workers are not equal per rank, I think it can lead to issues while distributing the data.

Reply:
good question! yes, we are assuming uniform distribution, but let me think, if we can make it more agile and support  uneven distribution